zerocopy-derive/src/lib.rs

```diff
@@ -610,77 +564,37 @@ fn impl_block<D: DataExt>(
         GenericParam::Const(cnst) => quote!(#cnst),
     });
 
-    if require_trait_bound {
-        for ty in non_type_param_field_types {
-            where_clause.predicates.push(parse_quote!(#ty: zerocopy::#trait_ident));
-        }
-    }
-
-    match (field_types.is_empty(), padding_check) {
-        (true, _) | (false, PaddingCheck::None) => (),
-        (false, PaddingCheck::Struct) => {
-            let fields = field_types.iter();
-            // `parse_quote!` doesn't parse macro invocations in const generics
-            // properly without enabling syn's `full` feature, so the type has
-            // to be manually constructed as `syn::Type::Verbatim`.
-            //
-            // This where clause is equivalent to adding:
-            // ```
-            // HasPadding<Foo, {struct_has_padding!(Foo, a, b, ...)}>: ShouldBe<false>
-            // ```
-            // with fully-qualified paths.
-            where_clause.predicates.push(syn::WherePredicate::Type(syn::PredicateType {
-                lifetimes: None,
-                bounded_ty: syn::Type::Verbatim(quote!(zerocopy::derive_util::HasPadding<#type_ident, {zerocopy::struct_has_padding!(#type_ident, #(#fields),*)}>)),
-                colon_token: syn::Token![:](Span::mixed_site()),
-                bounds: parse_quote!(zerocopy::derive_util::ShouldBe<false>),
-            }));
-        }
-        (false, PaddingCheck::Union) => {
-            let fields = field_types.iter();
-            // `parse_quote!` doesn't parse macro invocations in const generics
-            // properly without enabling syn's `full` feature, so the type has
-            // to be manually constructed as `syn::Type::Verbatim`.
-            //
-            // This where clause is equivalent to adding:
-            // ```
-            // HasPadding<Foo, {union_has_padding!(Foo, a, b, ...)}>: ShouldBe<false>
-            // ```
-            // with fully-qualified paths.
-            where_clause.predicates.push(syn::WherePredicate::Type(syn::PredicateType {
-                lifetimes: None,
-                bounded_ty: syn::Type::Verbatim(quote!(zerocopy::derive_util::HasPadding<#type_ident, {zerocopy::union_has_padding!(#type_ident, #(#fields),*)}>)),
-                colon_token: syn::Token![:](Span::mixed_site()),
-                bounds: parse_quote!(zerocopy::derive_util::ShouldBe<false>),
-            }));
-        }
-    }
-
-    // We use a constant to force the compiler to emit an error when a concrete
-    // type does not satisfy the where clauses on its impl.
-    let use_concrete = if input.generics.params.is_empty() {
-        Some(quote! {
-            const _: () = {
-                fn must_implement_trait<T: zerocopy::#trait_ident + ?::core::marker::Sized>() {}
-                let _ = must_implement_trait::<#type_ident>;
-            };
-        })
-    } else {
-        None
-    };
```

Comment:
@djkoloski Calling this out: This is the only change here that affects the generated code. It doesn't seem to be necessary now that we add bounds for field types to the `where` clause, but I wanted to make sure to get scrutiny on this. I'm already surprised that we can put bounds on non-generic field types (like `u8`) in the `where` clause because I would assume that requires trivial_bounds ([which is unstable](https://github.com/rust-lang/rust/issues/48214)), but it does work:

```rust
trait Foo {}
impl Foo for u8 {}
impl Foo for () where u8: Foo {} // not trivial_bounds apparently??
```

Reply:
Discussed offline: We're not sure why this is valid, but it's definitely valid. It's been compiling successfully since 1.6.